src/libmongoc/tests/unified/entity-map.c

```diff
@@ -718,3 +858,382 @@ entity_map_get (entity_map_t *entity_map, const char *id, bson_error_t *error)
+   lsid = entity_map_get_lsid (em, (char *) id, error);
+   if (!lsid) {
+      goto done;
+   }
+
+   session_val = bson_val_from_bson (lsid);
+   if (!bson_matcher_match (matcher, session_val, actual, path, error)) {
+      goto done;
+   }
+
+
+   ret = true;
+done:
+   bson_val_destroy (session_val);
+   return ret;
+}
+
+/* implement $$matchesEntity */
+bool
+special_matches_entity (bson_matcher_t *matcher,
+                        const bson_t *assertion,
+                        const bson_val_t *actual,
+                        void *ctx,
+                        const char *path,
+                        bson_error_t *error)
+{
+   bool ret = false;
+   bson_iter_t iter;
+   entity_map_t *em = (entity_map_t *) ctx;
+   bson_val_t *entity_val = NULL;
+   const char *id;
+
+   bson_iter_init (&iter, assertion);
+   BSON_ASSERT (bson_iter_next (&iter));
+
+   if (!BSON_ITER_HOLDS_UTF8 (&iter)) {
+      test_set_error (error,
+                      "unexpected $$matchesEntity does not contain utf8: %s",
+                      tmp_json (assertion));
+      goto done;
+   }
+
+   id = bson_iter_utf8 (&iter, NULL);
+   entity_val = entity_map_get_bson (em, id, error);
+   if (!entity_val) {
+      goto done;
+   }
+
+   if (!bson_matcher_match (matcher, entity_val, actual, path, error)) {
+      goto done;
+   }
+
+   ret = true;
+done:
+   return ret;
+}
+
+bool
+entity_map_match (entity_map_t *em,
+                  const bson_val_t *expected,
+                  const bson_val_t *actual,
+                  bson_error_t *error)
+{
+   bson_matcher_t *matcher;
+   bool ret;
+
+   matcher = bson_matcher_new ();
+   bson_matcher_add_special (
+      matcher, "$$sessionLsid", special_session_lsid, em);
+   bson_matcher_add_special (
+      matcher, "$$matchesEntity", special_matches_entity, em);
+   ret = bson_matcher_match (matcher, expected, actual, "", error);
+   bson_matcher_destroy (matcher);
+   return ret;
+}
+
+char *
+event_list_to_string (event_t *events)
+{
+   bson_string_t *str = NULL;
+   event_t *eiter = NULL;
+
+   str = bson_string_new ("");
+   LL_FOREACH (events, eiter)
+   {
+      bson_string_append_printf (str, "- %s:", eiter->type);
+      if (eiter->command_name) {
+         bson_string_append_printf (str, " cmd=%s", eiter->command_name);
+      }
+      if (eiter->database_name) {
+         bson_string_append_printf (str, " db=%s", eiter->database_name);
+      }
+      if (eiter->command) {
+         bson_string_append_printf (str, " sent %s", tmp_json (eiter->command));
+      }
+      if (eiter->reply) {
+         bson_string_append_printf (
+            str, " received %s", tmp_json (eiter->reply));
+      }
+      bson_string_append (str, "\n");
+   }
+   return bson_string_free (str, false);
+}
+
+
+bool
+entity_map_end_session (entity_map_t *em, char *session_id, bson_error_t *error)
+{
+   bool ret = false;
+   entity_t *entity = NULL;
+
+   entity = entity_map_get (em, session_id, error);
```

Comment:
So because we use `entity_map_get` and not `entity_map_get_session` this function is safe to call on sessions that have already ended?